### PR TITLE
fix: hover and focus openable navigation menu

### LINKED ISSUE #1999: useHover sets focus on reference, causing focus indicator to appear
Reproduction: https://codesandbox.io/s/cool-framework-yzkhwh?file=/src/DropdownMenu.tsx

Video just in case:
https://user-images.githubusercontent.com/1892091/205591138-78b7eaa4-1cfa-4c7d-8722-1c08acfda334.mp4

In the video, in the beginning hovering the reference button works as expected, e.g. there is no visible focus indicator. Once I tab to the button at least once and close the menu, after that when hovering the button the focus indicator is shown.

Note that sometimes it does not even need the tabbing, it just shows the focus indicator when hovering.

This also does not seem to be related to the use of useFocus hook, i.e. if I only use useHover, it still behaves like this.

### PR DESCRIPTION
**Pull request by mihkeleidast**
Potentially resolves #1999.

### CODE REVIEW

**Comment by atomiks** [comment]
Prior to v0.13.0 it did this but it has some issues like on Safari because buttons don't receive focus when pressing them by default :\ so focus gets returned to nothing when pressing `esc`. There are also other edge cases where it's preferrable to use the reference first.

**Comment by mihkeleidast** [comment]
Good old Safari! I'm not sure of the details, as there aren't any tests covering this nor the other edgecases. But as we know when the `esc` key usecase specifically happens in `onDismiss`, could we overwrite `previouslyFocusedElement` in that scenario to be the reference?

**Comment by atomiks** [comment]
I guess that could work

**Comment by mihkeleidast** [comment]
Alright, did that, plus added two tests so this specific behavior would not be broken:
- that the reference is not focused when only hovering/unhovering it
- that the reference is focused when opening via hover, but closing via esc. I think this should also cover the Safari edgecase you mentioned.

Any other test (edge)cases that would be good to cover here?

**Comment by mihkeleidast** [comment]
One other thing I noticed was that if I tab into the floating element, then close it via esc, the floating element actually re-opens because of the reference getting focused. I see that useFocus does look at the dismiss event and the onFocus open handler depends on that, but only for `referencePress` at the moment. Is there a reason for that, or should it also not re-open when it is closed via ESC? https://github.com/floating-ui/floating-ui/blob/master/packages/react-dom-interactions/src/hooks/useFocus.ts#L61

**Comment by atomiks** [comment]
That could/should be added as well yeah

File: packages/react-dom-interactions/src/FloatingFocusManager.tsx [hunk 1]
```diff
@@ -345,6 +344,10 @@ export function FloatingFocusManager<RT extends ReferenceType = ReferenceType>({
     // Dismissing via outside press should always ignore `returnFocus` to
     // prevent unwanted scrolling.
     function onDismiss(payload: DismissPayload) {
+      if (payload.type === 'escapeKey') {
+        previouslyFocusedElementRef.current = refs.domReference.current;
+      }
```
Comment (atomiks) [inline packages/react-dom-interactions/src/FloatingFocusManager.tsx:349]: If the user presses an explicit close button it should probably still return focus to the reference though in Safari - it will return focus to nothing here?

Comment (atomiks) [reply packages/react-dom-interactions/src/FloatingFocusManager.tsx:349]: I think your line here handles it 👍 

```js
      if (contains(floating, activeElement(doc))) {
        previouslyFocusedElementRef.current = refs.domReference.current;
      }
```

File: packages/react-dom-interactions/src/FloatingFocusManager.tsx [hunk 2]
```diff
@@ -364,12 +367,16 @@ export function FloatingFocusManager<RT extends ReferenceType = ReferenceType>({
     return () => {
       events.off('dismiss', onDismiss);
 
+      if (contains(floating, activeElement(doc))) {
+        previouslyFocusedElementRef.current = refs.domReference.current;
+      }
```
[the remark this answers is not in the document]
Comment (atomiks) [reply packages/react-dom-interactions/src/FloatingFocusManager.tsx:372]: Actually this applies to all places where you're setting it — as the DOM reference is optional (e.g. dialog that opens by itself)

[the remark this answers is not in the document]
Comment (mihkeleidast) [reply packages/react-dom-interactions/src/FloatingFocusManager.tsx]: Good point, added a check!

**Comment by atomiks** [comment]
Thanks!
